drv/lpc55-rng/src/main.rs

```diff
@@ -9,119 +9,124 @@
-        Lpc55Rng(BlockRng::new(Lpc55Core::new()))
-    }
-
-    fn init(&self) {
-        self.0
-            .core
-            .pmc
-            .pdruncfg0
-            .modify(|_, w| w.pden_rng().poweredon());
-
-        self.0.core.syscon.enable_clock(Peripheral::Rng);
+task_slot!(SYSCON, syscon_driver);
 
-        self.0.core.syscon.enter_reset(Peripheral::Rng);
-        self.0.core.syscon.leave_reset(Peripheral::Rng);
-    }
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    None,
+    #[cfg(feature = "dice-seed")]
+    NoDiceSeed,
+    #[cfg(feature = "dice-seed")]
+    HandoffError(HandoffDataLoadError),
+    #[cfg(feature = "dice-seed")]
+    NoSeedPersonalization,
 }
 
-impl RngCore for Lpc55Rng {
-    fn next_u32(&mut self) -> u32 {
-        self.0.next_u32()
-    }
-    fn next_u64(&mut self) -> u64 {
-        self.0.next_u64()
-    }
-    fn fill_bytes(&mut self, bytes: &mut [u8]) {
-        self.0.fill_bytes(bytes)
-    }
-    fn try_fill_bytes(&mut self, bytes: &mut [u8]) -> Result<(), Error> {
-        if self.0.core.pmc.pdruncfg0.read().pden_rng().bits() {
-            return Err(RngError::PoweredOff.into());
-        }
-
-        self.0.try_fill_bytes(bytes)
-    }
-}
+ringbuf!(Trace, 16, Trace::None);
 
 // low-budget rand::rngs::adapter::ReseedingRng w/o fork stuff
-struct ReseedingRng<T: SeedableRng> {
+struct ReseedingRng<T: SeedableRng, R: RngCore, H: Digest> {
     inner: T,
-    reseeder: Lpc55Rng,
+    reseeder: R,
     threshold: usize,
     bytes_until_reseed: usize,
+    mixer: H,
 }
 
-impl<T> ReseedingRng<T>
+impl<T, R, H> ReseedingRng<T, R, H>
 where
-    T: SeedableRng,
+    T: SeedableRng<Seed = [u8; 32]> + RngCore,
+    R: RngCore,
+    H: FixedOutputReset + Default + Digest,
+    [u8; 32]: From<GenericArray<u8, <H as OutputSizeUser>::OutputSize>>,
 {
-    fn new(mut reseeder: Lpc55Rng, threshold: usize) -> Result<Self, Error> {
-        use ::core::usize::MAX;
+    fn new(
+        seed: Option<&RngSeed>,
+        mut reseeder: R,
+        pid: Option<&[u8; SUBJECT_CN_LENGTH]>,
+        threshold: usize,
+    ) -> Result<Self, Error> {
+        let threshold = if threshold == 0 {
+            usize::MAX
+        } else {
+            threshold
+        };
 
-        let threshold = if threshold == 0 { MAX } else { threshold };
+        let mut mixer = H::default();
+        if let Some(seed) = seed {
+            // mix platform unique seed derived by measured boot
+            Digest::update(&mut mixer, seed.as_bytes());
+        }
+
+        if let Some(pid) = pid {
+            // mix in unique platform id
+            Digest::update(&mut mixer, pid);
+        }
+
+        // w/ 32 bytes from HRNG
+        let mut buf = Zeroizing::new(T::Seed::default());
+        reseeder.try_fill_bytes(buf.as_mut())?;
+        Digest::update(&mut mixer, buf.as_ref());
+
+        // create initial instance of the SeedableRng from the seed
+        let inner = T::from_seed(mixer.finalize_fixed_reset().into());
```

Comment:
FYI, you might see lighter stack usage here by using one of the `finalize_into` operations, instead of returning the array. It might also not change anything. Hard to tell.